src/Microsoft.DotNet.Wpf/src/System.Windows.Primitives/Windows/Win32/Foundation/ComScope.cs

```diff
@@ -0,0 +1,102 @@
+///  </para>
+/// </remarks>
+/// <typeparam name="T">
+///  This should be one of the struct COM definitions as generated by CsWin32. Ideally we'd constrain to
+///  <see cref="IUnknown.Interface"/> or some other interface tag to enforce that this is being used around
+///  a struct that is actually a COM wrapper.
+/// </typeparam>
+internal readonly unsafe ref struct ComScope<T> where T : unmanaged, IComIID
+{
+    // Keeping internal as nint allows us to use Unsafe methods to get significantly better generated code.
+    private readonly nint _value;
+    public T* Value => (T*)_value;
+    public IUnknown* AsUnknown => (IUnknown*)_value;
+
+    public ComScope(T* value) => _value = (nint)value;
+
+    // Can't add an operator for IUnknown* as we have ComScope<IUnknown>.
+
+    public static implicit operator T*(in ComScope<T> scope) => (T*)scope._value;
+
+    public static implicit operator void*(in ComScope<T> scope) => (void*)scope._value;
+
+    public static implicit operator nint(in ComScope<T> scope) => scope._value;
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static implicit operator T**(in ComScope<T> scope) => (T**)Unsafe.AsPointer(ref Unsafe.AsRef(scope._value));
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static implicit operator void**(in ComScope<T> scope) => (void**)Unsafe.AsPointer(ref Unsafe.AsRef(scope._value));
+
+    public bool IsNull => _value == 0;
+
+    public ComScope<TTo> TryQuery<TTo>(out HRESULT hr) where TTo : unmanaged, IComIID
+    {
+        ComScope<TTo> scope = new(null);
+        hr = ((IUnknown*)Value)->QueryInterface(IID.Get<TTo>(), scope);
+        return scope;
+    }
+
+    /// <summary>
+    ///  Attempt to create a <see cref="ComScope{T}"/> from the given COM interface.
+    /// </summary>
+    public static ComScope<T> TryQueryFrom<TFrom>(TFrom* from, out HRESULT hr) where TFrom : unmanaged, IComIID
+    {
+        ComScope<T> scope = new(null);
+        hr = from is null ? HRESULT.E_POINTER : ((IUnknown*)from)->QueryInterface(IID.Get<T>(), scope);
+        return scope;
+    }
+
+    /// <summary>
+    ///  Create a <see cref="ComScope{T}"/> from the given COM interface. Throws on failure.
+    /// </summary>
+    public static ComScope<T> QueryFrom<TFrom>(TFrom* from) where TFrom : unmanaged, IComIID
+    {
+        if (from is null)
+        {
+            HRESULT.E_POINTER.ThrowOnFailure();
+        }
+
+        ComScope<T> scope = new(null);
+        ((IUnknown*)from)->QueryInterface(IID.Get<T>(), scope).ThrowOnFailure();
+        return scope;
+    }
+
+    public void Dispose()
+    {
+        IUnknown* unknown = (IUnknown*)_value;
+
+        // Really want this to be null after disposal to avoid double releases, but we also want
+        // to maintain the readonly state of the struct to allow passing as `in` without creating implicit
+        // copies (which would break the T** and void** operators).
```

Comment:
> but we also want to maintain the readonly state of the struct to allow passing as `in` without creating implicit copies

😕 `in` will still work without making implicit copies, even if the type is not readonly. The defensive copy for a parameter passed as `in` would occur at the time a non-readonly member is accessed (of which this type does not have any), except for `Dispose()` which would not be allowed.

Reply:
We're deliberately clearing `_value` here which is very much readonly.